asim: rework metrics

kvoli opened the pull request:
This series of commits reworks the metrics component in asim to decouple tracking and writing. This patch also adds the ability to track per-store metrics.

part of https://github.com/cockroachdb/cockroach/issues/83990

Release note: None

kvoli committed:
asim: move test workload gen into workload pkg

Previously the testing utility to generate a simple workload existed
within the `asim` pkg. This patch moves it into the `workload` pkg so
that it can be depended upon by additional pkgs, such as metrics.

Release note: None

kvoli committed:
asim: move metrics_*.go into metrics pkg

This patch moves `metrics_tracker.go` and `metrics.go` into a new pkg
`metrics`. This is done to decouple the metrics code from the `asim`
pkg.

Release note: None

kvoli committed:
asim: refactor metrics to use listener

Previously asim metrics were tracked and written using
`ClusterMetricsTracker`, this patch decouples tracking and writing by
introducing a listener like pattern.

A single `Tracker` exists per simulation. Multiple different listeners,
which implement the `StoreMetricsListener` interface can be registered
against the tracker at initialization. The listeners then receive
a `StoreMetrics` for every store, on each tick.

Release note: None

kvoli committed:
asim: add store metrics tracker

Previously, only aggregate cluster metrics were tracked during
simulation. This patch adds a metrics tracker for per-store metrics.
This patch enables future plotting of per-store timeseries, over a
simulation run and additional debugging ability.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/93951)
<!-- Reviewable:end -->

kvoli committed:
asim: [wip] add metric plotter

This commit adds a metric plotter that outputs ascii series graphs. The
purpose of this addition is for visual feedback on metrics gathered
during simulation.

Release note: None

kvoli committed:
asim: tick workload in foreground

Previously, the workload generator was ticked every background interval
tick to reduce by enabling greater batching. This is problematic however
when the workload ticks infrequently and the values are rated over a
short period. An example problem is that it can lead to extremely high
rated values for queries-per-second when a batch is applied for 10s of
load in only a single tick.

Release note: None

kvoli committed:
asim: use replica load in place of replicastats

Previously, the allocator simulator used replica stats to track the
rated stats of a replica. Replica load is a group of stats objects that
is better suited. This patch replaces replica stats with replica load.

Release note: None

kvoli committed:
asim: update local store after lease transfer

Previously, the simulator relied on a bug where gossip shared a
reference to each store descriptor in the cluster.

This was fixed in #93945. This commit updates the simulator code to
update the storepool state like the real code.

Release note: None

kvoli commented:
Going to defer the metrics changes to after the later fixes in this PR which I'll split out into a new PR.

kvoli committed:
asim: [wip] init state clock with start time

Previously, the clock that is shared among all simulated stores was
initialized to zero. The clock would then jump to the start time after
simulator ticking begun. This is problematic for rated components such
as replica stats which rely on the clock. This commit initializes the
state clock with a passed in argument defining the start time.

Release note: None

kvoli commented:
subsumed by https://github.com/cockroachdb/cockroach/pull/96822/commits
